SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp

```diff
@@ -0,0 +1,196 @@
+  }
+};
+
+// Descriptor class for the case of calling constructor in rvalue in an
+// expression context.
+struct rval_in_expr {
+  static std::string get_description() { return "rvalue in an expression"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(DstT ref_data, SrcT *const out) {
+    esimd::simd<SrcT, NumElems> simd_by_rval;
+    simd_by_rval = esimd::simd<SrcT, NumElems>(ref_data);
+    simd_by_rval.copy_to(out);
+  }
+};
+
+// Descriptor class for the case of calling constructor in const reference
+// context.
+class const_ref {
+public:
+  static std::string get_description() { return "const reference"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(DstT ref_data, SrcT *const out) {
+    call_simd_by_const_ref<SrcT, NumElems>(
+        esimd::simd<SrcT, NumElems>(ref_data), out);
+  }
+
+private:
+  template <typename SrcT, typename DstT, int NumElems>
+  static void
+  call_simd_by_const_ref(const esimd::simd<SrcT, NumElems> &simd_by_const_ref,
+                         SrcT *const out) {
+    simd_by_const_ref.copy_to(out);
+  }
+};
+
+template <typename DstT, int NumElems, typename ContextT>
+class ConvCtorTestDescription : public ITestDescription {
+public:
+  ConvCtorTestDescription(size_t index, DstT retrieved_val, DstT expected_val,
+                          const std::string &src_data_type,
+                          const std::string &dst_data_type)
+      : m_src_data_type(src_data_type), m_dst_data_type(dst_data_type),
+        m_retrieved_val(retrieved_val), m_expected_val(expected_val),
+        m_index(index) {}
+
+  std::string to_string() const override {
+    // TODO: Make strings for fp values more short during failure output, may be
+    // by using hex representation
+    std::string log_msg("Failed for simd<");
+
+    log_msg += m_src_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", with context: " + ContextT::get_description();
+    log_msg += ", with value type: " + m_dst_data_type;
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", at index: " + std::to_string(m_index);
+
+    return log_msg;
+  }
+
+private:
+  const std::string m_src_data_type;
+  const std::string m_dst_data_type;
+  const DstT m_retrieved_val;
+  const DstT m_expected_val;
+  const size_t m_index;
+};
+
+template <typename, int, typename> class Kernel;
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename SrcT, int NumElems, typename DstT, typename TestCaseT>
+class run_test {
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &src_data_type,
+                  const std::string &dst_data_type) {
+    bool passed = true;
+    const std::vector<SrcT> ref_data =
+        generate_converted_ref_data<SrcT, DstT, 1>();
```

Comment:
```suggestion
        generate_ref_conv_data<SrcT, DstT, 1>();
```
or
```suggestion
        generate_ref_data_for_conversion<SrcT, DstT, 1>();
```

Reply:
done. `generate_ref_conv_data` the current name